flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -321,4 +328,37 @@ private void registerReader() {
 	ListState<SplitT> getReaderState() {
 		return readerState;
 	}
+
+	// ----------- private class --------------
+
+	private static class CountingDataOutput<T> implements DataOutput<T> {
```

Comment:
Can we put the counting in `SourceOperatorStreamTask`?
That way we can
  - avoid yet another wrapping layer for the output
  - have the record counting the responsibility of the task and operator chain, which looks like the level on which it is solved in the remaining cases